x-pack/auditbeat/module/system/user/users_linux.go

```diff
@@ -68,6 +85,51 @@ func GetUsers() (users []*User, err error) {
 	return users, nil
 }
 
+// readGroupFile reads /etc/group and returns two maps:
+// The first maps group IDs to groups.
+// The second maps group members (user names) to groups.
+// See getgrent(3) for details of the structs.
+func readGroupFile() (map[uint32]Group, map[string][]Group, error) {
+	C.setgrent()
+	defer C.endgrent()
+
+	groupIdMap := make(map[uint32]Group)
+	groupMemberMap := make(map[string][]Group)
+	for cgroup, err := C.getgrent(); cgroup != nil; cgroup, err = C.getgrent() {
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "error while reading group file")
+		}
+
+		groupName := C.GoString(cgroup.gr_name)
+		gid := uint32(cgroup.gr_gid)
+
+		group := Group{
+			Name: groupName,
+			GID:  gid,
+		}
+
+		groupIdMap[gid] = group
+
+		/*
+			group.gr_mem is a NULL-terminated array of pointers to user names (char **)
+			which makes some pointer arithmetic necessary to read it.
+		*/
+		for i := 0; ; i++ {
+			offset := (unsafe.Sizeof(unsafe.Pointer(*cgroup.gr_mem)) * uintptr(i))
+			member := *(**C.char)(unsafe.Pointer(uintptr(unsafe.Pointer(cgroup.gr_mem)) + offset))
```

Comment:
Tough one, but LGTM from what I can tell. 👍

Reply:
Yeah, it's not pretty.